### PR TITLE
Augment Python PlanningSceneInterface API

### LINKED ISSUE #3527: Allowed Collision Matrix not updating using apply_planning_scene service.
### Description

I am attempting to modify the Allowed Collision Matrix (ACM) to prevent collision checking with an object added to the scene.
I have followed the steps for updating the ACM recommended here: [Updating allowed collision matrix used for motion plans](https://github.com/ros-planning/moveit/issues/2818)

I retrieve the ACM using the `get_planning_scene` service, modify the ACM, and send it to the `apply_planning_scene` service.
However, checking `get_planning_scene` after shows the update has not been applied.
Notably I am able to successfully add new objects to the scene using `apply_planning_scene`, but the ACM is not changed.
The same is true if I use the `planning_scene` topic.

### Your environment
* ROS Distro: Noetic
* OS Version:  Ubuntu 20.04
* Binary
* 1.1.13-2focal.20231014.010446

### Steps to reproduce
The following code is a minimal test intended to replicate the issue.
Run it alongside MoveIt.

```
import rospy
import moveit_commander
from moveit_msgs.srv import GetPlanningScene, ApplyPlanningSceneRequest
from moveit_msgs.msg import PlanningScene, AllowedCollisionEntry


def allow_collisions_with_object(obj_name, scene):
    """ Updates the MoveIt PlanningScene using the AllowedCollisionMatrix to ignore collisions for an object """
    # Set up service to get the current planning scene
    service_timeout = 5.0
    _get_planning_scene = rospy.ServiceProxy(
        "get_planning_scene", GetPlanningScene
    )
    _get_planning_scene.wait_for_service(service_timeout)

    request = GetPlanningScene()
    request.components = 0    # Get just the Allowed Collision Matrix
    planning_scene = _get_planning_scene.call(request)
    print(f"\n\n\n--- allowed_collision_matrix before update:{planning_scene.scene.allowed_collision_matrix} ---\n\n\n")

    # Set this object to ignore collisions with all objects. The entry values are not updated
    planning_scene.scene.allowed_collision_matrix.entry_names.append(obj_name)
    for entry in planning_scene.scene.allowed_collision_matrix.entry_values:
        entry.enabled.append(False)
    enabled = [False for i in range(len(planning_scene.scene.allowed_collision_matrix.entry_names))]
    entry = AllowedCollisionEntry(enabled=enabled)  # Test kwarg in constructor
    planning_scene.scene.allowed_collision_matrix.entry_values.append(entry)

    # Set the default entries. They are also not updated
    planning_scene.scene.allowed_collision_matrix.default_entry_names = [obj_name]
    planning_scene.scene.allowed_collision_matrix.default_entry_values = [False]
    planning_scene.scene.is_diff = True # Mark this as a diff message to force an update of the allowed collision matrix
    planning_scene.scene.robot_state.is_diff = True

    planning_scene_diff_req = ApplyPlanningSceneRequest()
    planning_scene_diff_req.scene = planning_scene.scene

    # Updating the Allowed Collision Matrix through the apply_planning_scene service shows no effect.
    # However, adding objects to the planning scene works fine.
    # scene._apply_planning_scene_diff.call(planning_scene_diff_req)
    scene.apply_planning_scene(planning_scene.scene)

    # Attempting to use the planning_scene topic for asynchronous updates also does not work
    # planning_scene_pub = rospy.Publisher("planning_scene", PlanningScene, queue_size=5)
    # planning_scene_pub.publish(planning_scene.scene)

    print(f"\n--- Sent message:{planning_scene.scene.allowed_collision_matrix} ---\n")

    # The planning scene retrieved after the update should have taken place shows the Allowed Collision Matrix is the same as before
    request = GetPlanningScene()
    request.components = 0    # Get just the Allowed Collision Matrix
    planning_scene = _get_planning_scene.call(request)
    print(f"\n--- allowed_collision_matrix after update:{planning_scene.scene.allowed_collision_matrix} ---\n")


if __name__ == "__main__":
    # Moveit scene variable
    scene = moveit_commander.PlanningSceneInterface()
    allow_collisions_with_object("temp_obj_name", scene)
    
```

### Expected behaviour
After the update, the ACM is retrieved. It should now show the changes in the ACM requested in the message.

### Actual behaviour
After the update, the ACM is the same as it was before, with none of our changes to the default entries or entries displayed.

### Backtrace or Console output
```
--- allowed_collision_matrix before update:entry_names: 
  - Gripper_cam_link
  - Gripper_link
  - Link1
  - Link2
  - Link3
  - Link4
  - Link5
  - Link6
  - Mount1
  - base_link
  - robot_body_link
entry_values: 
  - 
    enabled: [False, True, False, False, True, True, True, True, True, False, False]
  - 
    enabled: [True, False, False, False, True, True, True, True, True, False, False]
  - 
    enabled: [False, False, False, True, True, True, False, False, False, True, True]
  - 
    enabled: [False, False, True, False, True, True, False, False, False, True, True]
  - 
    enabled: [True, True, True, True, False, True, True, True, True, True, False]
  - 
    enabled: [True, True, True, True, True, False, True, True, True, False, False]
  - 
    enabled: [True, True, False, False, True, True, False, True, True, False, False]
  - 
    enabled: [True, True, False, False, True, True, True, False, True, False, False]
  - 
    enabled: [True, True, False, False, True, True, True, True, False, False, False]
  - 
    enabled: [False, False, True, True, True, False, False, False, False, False, True]
  - 
    enabled: [False, False, True, True, False, False, False, False, False, True, False]
default_entry_names: []
default_entry_values: [] ---


--- Sent message:entry_names: 
  - Gripper_cam_link
  - Gripper_link
  - Link1
  - Link2
  - Link3
  - Link4
  - Link5
  - Link6
  - Mount1
  - base_link
  - robot_body_link
  - temp_obj_name
entry_values: 
  - 
    enabled: [False, True, False, False, True, True, True, True, True, False, False, False]
  - 
    enabled: [True, False, False, False, True, True, True, True, True, False, False, False]
  - 
    enabled: [False, False, False, True, True, True, False, False, False, True, True, False]
  - 
    enabled: [False, False, True, False, True, True, False, False, False, True, True, False]
  - 
    enabled: [True, True, True, True, False, True, True, True, True, True, False, False]
  - 
    enabled: [True, True, True, True, True, False, True, True, True, False, False, False]
  - 
    enabled: [True, True, False, False, True, True, False, True, True, False, False, False]
  - 
    enabled: [True, True, False, False, True, True, True, False, True, False, False, False]
  - 
    enabled: [True, True, False, False, True, True, True, True, False, False, False, False]
  - 
    enabled: [False, False, True, True, True, False, False, False, False, False, True, False]
  - 
    enabled: [False, False, True, True, False, False, False, False, False, True, False, False]
  - 
    enabled: [False, False, False, False, False, False, False, False, False, False, False, False]
default_entry_names: 
  - temp_obj_name
default_entry_values: [False] ---


--- allowed_collision_matrix after update:entry_names: 
  - Gripper_cam_link
  - Gripper_link
  - Link1
  - Link2
  - Link3
  - Link4
  - Link5
  - Link6
  - Mount1
  - base_link
  - robot_body_link
entry_values: 
  - 
    enabled: [False, True, False, False, True, True, True, True, True, False, False]
  - 
    enabled: [True, False, False, False, True, True, True, True, True, False, False]
  - 
    enabled: [False, False, False, True, True, True, False, False, False, True, True]
  - 
    enabled: [False, False, True, False, True, True, False, False, False, True, True]
  - 
    enabled: [True, True, True, True, False, True, True, True, True, True, False]
  - 
    enabled: [True, True, True, True, True, False, True, True, True, False, False]
  - 
    enabled: [True, True, False, False, True, True, False, True, True, False, False]
  - 
    enabled: [True, True, False, False, True, True, True, False, True, False, False]
  - 
    enabled: [True, True, False, False, True, True, True, True, False, False, False]
  - 
    enabled: [False, False, True, True, True, False, False, False, False, False, True]
  - 
    enabled: [False, False, True, True, False, False, False, False, False, True, False]
default_entry_names: []
default_entry_values: [] ---
```

### PR DESCRIPTION
**Pull request by rhaschke**
This PR addresses several shortcomings of the current python API for PSI. See individual commit messages for details.
Closes #3527

### CODE REVIEW

File: moveit_commander/src/moveit_commander/planning_scene_interface.py
Comment (v4hn) [inline moveit_commander/src/moveit_commander/planning_scene_interface.py:141]:
```suggestion
        object.touch_links = touch_links if touch_links is not None else [link]
```

File: moveit_commander/src/moveit_commander/planning_scene_interface.py
Comment (v4hn) [inline moveit_commander/src/moveit_commander/planning_scene_interface.py:139]:
```suggestion
     def attach_object(self, object, link=None, touch_links=None):
        """Attach an object to the given link"""
        if isinstance(object, str):
            object = self.__make_existing(object)
        if isinstance(object, CollisionObject):
            object = AttachedCollisionObject(object=object)

    if link is not None:
        object.link_name = link
```

Looks like github does not support scattered diffs in the suggestions -.-
Anyway, I propose to support passing an `AttachedCollisionObject` which already specifies the link in the message.

File: moveit_commander/src/moveit_commander/planning_scene_interface.py
Comment (v4hn) [inline moveit_commander/src/moveit_commander/planning_scene_interface.py:143]: Why not get rid of the attach parameter here altogether? The current code will break with an `AttachedCollisionObject` and `attach == False` and vice versa, right? So we might as well check via `isinstance` as you do above.

File: moveit_ros/planning_interface/planning_scene_interface/include/moveit/planning_scene_interface/planning_scene_interface.h
Comment (v4hn) [inline moveit_ros/planning_interface/planning_scene_interface/include/moveit/planning_scene_interface/planning_scene_interface.h:126]:
```suggestion
  /** \brief Get given components from move_group's PlanningScene *
    * \param components bitfield as specified in moveit_msgs::PlanningSceneComponents */
```

File: moveit_ros/planning_interface/planning_scene_interface/src/planning_scene_interface.cpp
Comment (v4hn) [inline moveit_ros/planning_interface/planning_scene_interface/src/planning_scene_interface.cpp:238]:
```suggestion
      ROS_ERROR_STREAM_NAMED(LOGNAME, "Could not call planning scene service '" << planning_scene_service_.getService() << "'");
```

File: moveit_commander/src/moveit_commander/planning_scene_interface.py
Comment (v4hn) [inline moveit_commander/src/moveit_commander/planning_scene_interface.py:418]: Can you please comment on how usual/reasonable you think these monkey patches are?
I've never seen anyone patch ROS message objects before (granted, I didn't look for it) and at first glance it looks useful, but I also think it's a can of worms I'd prefer closed...

**Review by v4hn — CHANGES_REQUESTED** [review]
Great simplifications and better functionality, yay.

I wrote all suggestions on the website, so I guess format issues and typos are to be expected.

File: moveit_commander/src/moveit_commander/planning_scene_interface.py
Comment (rhaschke) [reply moveit_commander/src/moveit_commander/planning_scene_interface.py:143]: :+1:

Comment (rhaschke) [reply moveit_commander/src/moveit_commander/planning_scene_interface.py:142]: The attach argument is needed for __submit to actually add an attached object. There is no type-checking in that function.

File: moveit_commander/src/moveit_commander/planning_scene_interface.py
Comment (rhaschke) [reply moveit_commander/src/moveit_commander/planning_scene_interface.py:417]: This simply augments the standard ROS-message class `AllowedCollisionMatrix` with two new functions.

**Comment by rhaschke** [comment]
I hope I addressed all your comments, @v4hn. Sorry for the cluttered replies to your comments. I used VS code for this and it didn't assign them to the original locations.

File: moveit_commander/src/moveit_commander/planning_scene_interface.py
Comment (v4hn) [reply moveit_commander/src/moveit_commander/planning_scene_interface.py:145]: Yes, there isn't. I rather suggested to add it there. I see why it wasn't there originally in the "never check types" spirit, but then providing the type as a boolean argument isn't much better...

File: moveit_commander/src/moveit_commander/planning_scene_interface.py
Comment (v4hn) [reply moveit_commander/src/moveit_commander/planning_scene_interface.py:420]: I know what it does. I'm not a fan of monkey patching, but of course the resulting interface looks nice when used.
I don't know how roundabout it would be to add it directly in moveit_msgs instead of here for example.

File: moveit_commander/src/moveit_commander/planning_scene_interface.py
Comment (rhaschke) [reply moveit_commander/src/moveit_commander/planning_scene_interface.py:145]: Please feel free to augment my PR according to your idea. I still don't get it in full detail and I cannot assess the consequences of this change right now. At least, it is an API change - but of a private function. So, that shouldn't pose a big problem.
Sorry, I'm still somewhat dizzy due to my illness...

File: moveit_commander/src/moveit_commander/planning_scene_interface.py
Comment (rhaschke) [reply moveit_commander/src/moveit_commander/planning_scene_interface.py:420]: I am not aware of a possibility to add methods directly in moveit_msgs. Of course, this would be even nicer. If you know an option to do so, please implement that in a follow-up PR.

**Comment by rhaschke** [comment]
@v4hn, is this ready to get merged? The additional changes you asked for, go far beyond this PR.

**Review by v4hn — APPROVED** [review]

**Comment by v4hn** [comment]
You are right. I still don't like the monkey patches, but they are definitely useful and I don't see a better way to provide the methods.

I'm not sure whether you want to clean up the commits before merging or merge as squash, so I'll leave it to you.

**Pull request merged**
